lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -81,8 +81,29 @@ def add_node_data(node)
     node.add_server_facts(@server_facts)
   end
 
+  # Inline file metadata for static catalogs
+  # Initially restricted to files sourced from codedir via puppet:/// uri.
+  def inline_metadata(catalog, environment)
+    catalog.resources.find_all { |res| res.type == "File" }.each do |resource|
+      next if resource[:ensure] == 'absent'
+
+      next unless source = resource[:source]
+      next unless source =~ /^puppet:/
+
+      file = resource.to_ral
+
+      if file.recurse?
+        # TODO: find and replace metadata for children recursively
+      else
+        # TODO: find and replace metadata for the file
+      end
+    end
+
+    catalog
+  end
```

Comment:
Update: filtering happens after `storeconfigs` (and caching?), so we probably want to inline metadata before that.